server/src-lib/Hasura/App.hs

```diff
@@ -275,14 +271,48 @@ runHGEServer ServeOptions{..} InitCtx{..} initTime = do
   let apiInitTime = realToFrac $ Clock.diffUTCTime finishTime initTime
   unLogger logger $
     mkGenericLog LevelInfo "server" $ StartupTimeInfo "starting API server" apiInitTime
+  let warpSettings = Warp.setPort soPort
+                     . Warp.setHost soHost
+                     . Warp.setGracefulShutdownTimeout (Just 30) -- 30s graceful shutdown
+                     . Warp.setInstallShutdownHandler (shutdownHandler logger shutdownApp eventEngineCtx _icPgPool)
+                     $ Warp.defaultSettings
   liftIO $ Warp.runSettings warpSettings app
 
   where
+    -- | prepareEvents is a function to unlock all the events that are
+    -- locked and unprocessed, which is called while hasura is started.
+    -- Locked and unprocessed events can occur in 2 ways
+    -- 1.
+    -- Hasura's shutdown was not graceful in which all the fetched
+    -- events will remain locked and unprocessed(TODO: clean shutdown)
+    -- state.
+    -- 2.
+    -- There is another hasura instance which is processing events and
+    -- it will lock events to process them.
+    -- So, unlocking all the locked events might re-deliver an event(due to #2).
     prepareEvents pool (Logger logger) = do
       liftIO $ logger $ mkGenericStrLog LevelInfo "event_triggers" "preparing data"
-      res <- runUnlockTx pool unlockAllEvents
+      res <- liftIO $ runTx pool (Q.ReadCommitted, Nothing) unlockAllEvents
       either printErrJExit return res
 
+    -- | shutdownEvents will be triggered when a graceful shutdown has been inititiated, it will
+    -- get the locked events from the event engine context and then it will unlock all those events.
+    -- It may happen that an event may be processed more than one time, an event that has been already
+    -- processed but not been marked as delivered in the db will be unlocked by `shutdownEvents`
+    -- and will be processed when the events are proccessed next time.
+    shutdownEvents :: Q.PGPool -> Logger Hasura -> EventEngineCtx -> IO ()
+    shutdownEvents pool (Logger logger) EventEngineCtx {..} = do
+      liftIO $ logger $ mkGenericStrLog LevelInfo "event_triggers" "unlocking events that are locked by the HGE"
```

Comment:
should this be `unstructured`? Is this a new log-type: `event_triggers`?

Reply:
Okay nvm. I think the backpressure changes has an `unstructured` kind that should probably also use `event_triggers`: https://github.com/hasura/graphql-engine/blob/master/server/src-lib/Hasura/Events/Lib.hs#L236